api/users_test.go

```diff
@@ -842,6 +842,16 @@ func TestChangePasswordHandler(t *testing.T) {
 
 	api, w, m := apiSetup()
 
+	m.DescribeUserPoolClientFunc = func(input *cognitoidentityprovider.DescribeUserPoolClientInput) (*cognitoidentityprovider.DescribeUserPoolClientOutput, error) {
+		tokenValidDays := int64(1)
+		userPoolClient := &cognitoidentityprovider.DescribeUserPoolClientOutput{
+			UserPoolClient: &cognitoidentityprovider.UserPoolClientType{
+				RefreshTokenValidity: &tokenValidDays,
+			},
+		}
+		return userPoolClient, nil
+	}
+
```

Comment:
Its would be good if we could test to ensure the `expirationTime` and `refreshTokenExpirationTime` keys are present in the response body; issue is, `202` is also returned for an invalid user which is a deliberate security measure. This makes is slightly awkward. But by rewriting the `Convey("RespondToAuthChallenge - check expected responses", t, func() {}` test step to include the assertions like:

```suggestion
	Convey("RespondToAuthChallenge - check expected responses", t, func() {
		respondToAuthChallengeTests := []struct {
			respondToAuthChallengeFunction func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error)
			assertions                     func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse)
		}{
			{
				// Cognito successful password change
				func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error) {
					return &cognitoidentityprovider.RespondToAuthChallengeOutput{
						AuthenticationResult: &cognitoidentityprovider.AuthenticationResultType{
							AccessToken:  &accessToken,
							ExpiresIn:    &expireLength,
							IdToken:      &idToken,
							RefreshToken: &refreshToken,
						},
					}, nil
				},
				func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse) {
					var responseBody map[string]interface{}
					_ = json.Unmarshal(successResponse.Body, &responseBody)
					So(successResponse, ShouldNotBeNil)
					So(successResponse.Status, ShouldEqual, http.StatusAccepted)
					So(errorResponse, ShouldBeNil)
					So(responseBody["expirationTime"], ShouldNotBeNil)
					So(responseBody["refreshTokenExpirationTime"], ShouldNotBeNil)
				},
			},
			{
				// Cognito internal error
				func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error) {
					awsErrCode := "InternalErrorException"
					awsErrMessage := "Something strange happened"
					awsOrigErr := errors.New(awsErrCode)
					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
					return nil, awsErr
				},
				func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse) {
					So(successResponse, ShouldBeNil)
					So(errorResponse.Status, ShouldEqual, http.StatusInternalServerError)
				},
			},
			{
				// Cognito invalid session
				func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error) {
					awsErrCode := "CodeMismatchException"
					awsErrMessage := "session invalid"
					awsOrigErr := errors.New(awsErrCode)
					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
					return nil, awsErr
				},
				func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse) {
					So(successResponse, ShouldBeNil)
					So(errorResponse.Status, ShouldEqual, http.StatusBadRequest)
				},
			},
			{
				// Cognito invalid password
				func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error) {
					awsErrCode := "InvalidPasswordException"
					awsErrMessage := "password invalid"
					awsOrigErr := errors.New(awsErrCode)
					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
					return nil, awsErr
				},
				func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse) {
					So(successResponse, ShouldBeNil)
					So(errorResponse.Status, ShouldEqual, http.StatusBadRequest)
				},
			},
			{
				// Cognito invalid user
				func(input *cognitoidentityprovider.RespondToAuthChallengeInput) (*cognitoidentityprovider.RespondToAuthChallengeOutput, error) {
					awsErrCode := "UserNotFoundException"
					awsErrMessage := "user not found"
					awsOrigErr := errors.New(awsErrCode)
					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
					return nil, awsErr
				},
				func(successResponse *models.SuccessResponse, errorResponse *models.ErrorResponse) {
					So(successResponse, ShouldNotBeNil)
					So(successResponse.Status, ShouldEqual, http.StatusAccepted)
					So(errorResponse, ShouldBeNil)
				},
			},
		}

		for _, tt := range respondToAuthChallengeTests {
			m.RespondToAuthChallengeFunc = tt.respondToAuthChallengeFunction

			postBody := map[string]interface{}{"type": models.NewPasswordRequiredType, "email": email, "password": password, "session": session}
			body, _ := json.Marshal(postBody)
			r := httptest.NewRequest(http.MethodPut, changePasswordEndPoint, bytes.NewReader(body))

			successResponse, errorResponse := api.ChangePasswordHandler(ctx, w, r)

			tt.assertions(successResponse, errorResponse)
		}
	})
}
```
That will give us what we want for the test.